### ISSUE TITLE
get-external-data.py errors

### ISSUE THREAD

**Issue by sevensamurai** [Overv/openstreetmap-tile-server#244]
It seems the get-external-data script depends on external/flaky URLs and that causes the import to fail
That later on causes errors like:
Postgis Plugin: ERROR:  relation "icesheet_outlines" does not exist

Any way to provide these external files or configuration as import argument? 
The openstreetmap-carto/external-data.yml file seems to be the place, thinking maybe passing that as argument?
(reference this https://help.openstreetmap.org/questions/75856/when-building-tile-server-can-not-download-shapefile)

stacktrace below: using latest version as today

+ sudo chown -R renderer: /home/renderer/src

+ sudo -u renderer python3 /home/renderer/src/openstreetmap-carto/scripts/get-external-data.py -c /home/renderer/src/openstreetmap-carto/external-data.yml -D /home/renderer/src/openstreetmap-carto/data

INFO:root:Checking table simplified_water_polygons

INFO:root:Checking table water_polygons

INFO:root:Checking table icesheet_polygons

INFO:root:Checking table icesheet_outlines

Traceback (most recent call last):

  File "/usr/local/lib/python3.8/dist-packages/urllib3/connectionpool.py", line 699, in urlopen

    httplib_response = self._make_request(

  File "/usr/local/lib/python3.8/dist-packages/urllib3/connectionpool.py", line 445, in _make_request

    six.raise_from(e, None)

  File "<string>", line 3, in raise_from

  File "/usr/local/lib/python3.8/dist-packages/urllib3/connectionpool.py", line 440, in _make_request

    httplib_response = conn.getresponse()

  File "/usr/lib/python3.8/http/client.py", line 1344, in getresponse

    response.begin()

  File "/usr/lib/python3.8/http/client.py", line 307, in begin

    version, status, reason = self._read_status()

  File "/usr/lib/python3.8/http/client.py", line 276, in _read_status

    raise RemoteDisconnected("Remote end closed connection without"

http.client.RemoteDisconnected: Remote end closed connection without response

During handling of the above exception, another exception occurred:

Traceback (most recent call last):

  File "/usr/local/lib/python3.8/dist-packages/requests/adapters.py", line 439, in send

    resp = conn.urlopen(

  File "/usr/local/lib/python3.8/dist-packages/urllib3/connectionpool.py", line 755, in urlopen

    retries = retries.increment(

  File "/usr/local/lib/python3.8/dist-packages/urllib3/util/retry.py", line 532, in increment

    raise six.reraise(type(error), error, _stacktrace)

  File "/usr/local/lib/python3.8/dist-packages/urllib3/packages/six.py", line 769, in reraise

    raise value.with_traceback(tb)

  File "/usr/local/lib/python3.8/dist-packages/urllib3/connectionpool.py", line 699, in urlopen

    httplib_response = self._make_request(

  File "/usr/local/lib/python3.8/dist-packages/urllib3/connectionpool.py", line 445, in _make_request

    six.raise_from(e, None)

  File "<string>", line 3, in raise_from

  File "/usr/local/lib/python3.8/dist-packages/urllib3/connectionpool.py", line 440, in _make_request

    httplib_response = conn.getresponse()

  File "/usr/lib/python3.8/http/client.py", line 1344, in getresponse

    response.begin()

  File "/usr/lib/python3.8/http/client.py", line 307, in begin

    version, status, reason = self._read_status()

  File "/usr/lib/python3.8/http/client.py", line 276, in _read_status

    raise RemoteDisconnected("Remote end closed connection without"

urllib3.exceptions.ProtocolError: ('Connection aborted.', RemoteDisconnected('Remote end closed connection without response'))

During handling of the above exception, another exception occurred:

Traceback (most recent call last):

  File "/home/renderer/src/openstreetmap-carto/scripts/get-external-data.py", line 294, in <module>

    main()

  File "/home/renderer/src/openstreetmap-carto/scripts/get-external-data.py", line 230, in main

    download = s.get(source["url"], headers=headers)

  File "/usr/local/lib/python3.8/dist-packages/requests/sessions.py", line 555, in get

    return self.request('GET', url, **kwargs)

  File "/usr/local/lib/python3.8/dist-packages/requests/sessions.py", line 542, in request

    resp = self.send(prep, **send_kwargs)

  File "/usr/local/lib/python3.8/dist-packages/requests/sessions.py", line 655, in send

    r = adapter.send(request, **kwargs)

  File "/usr/local/lib/python3.8/dist-packages/requests/adapters.py", line 498, in send

    raise ConnectionError(err, request=request)

requests.exceptions.ConnectionError: ('Connection aborted.', RemoteDisconnected('Remote end closed connection without response'))

+ touch /var/lib/mod_tile/planet-import-complete

+ service postgresql stop

 * Stopping PostgreSQL 12 database server

   ...done.

+ exit 0

**Comment by nielsole** [comment]
Both the get-external-data.py as well as the external-data.yaml is provided by upstream: https://github.com/gravitystorm/openstreetmap-carto/blob/master/external-data.yml
The servers are managed by https://osmdata.openstreetmap.de/ (https://www.fossgis.de/)

Let's see if this error occurs more frequently then we could take this somewhere upstream and suggest fixing the server availability and/or adding retries in get-external-data.py . 
I wonder whether get-externa-data could be safely retried by run.sh or whether that leads to inconsistent data.

**Comment by nielsole** [comment]
Sounds like the script is cleaning up after itself https://github.com/gravitystorm/openstreetmap-carto/blob/master/scripts/get-external-data.py#L219

so we could probably wrap it in a retry loop. As it also only downloads the missing files again, it should also not be too bad data usage-wise https://github.com/gravitystorm/openstreetmap-carto/blob/master/scripts/get-external-data.py#L230

**Comment by sevensamurai** [comment]
thanks for reaching out, as far as I can tell it happens a lot from our side so that's why I'm keep to use pre-downloaded external data. The retry may just hide the problem, but makes sense to have at least that option

**Comment by DigitalCop** [comment]
I worked my way around this error. We are behind a corporate FW. I spinned up an Nginx container in a Docker network. I copied the .zip files into the container in /usr/share/nginx/html/download. Then I copied the external-data.yml file from the container and changed the url to http://<container_name>/download/ for each zip file. Thne i ran the installation with --network name identical to the Nginx container.  I also mounted the local external-data.yml file with the container. Here are mye two scripts:

Install:
```bash
docker run \
    -v "$(pwd)"/maps/europe-latest.osm.pbf:/data.osm.pbf \
    -v "$(pwd)"/maps/europe.poly:/data.poly \
    -v "$(pwd)"/data:/var/lib/postgresql/12/main \
    --mount type=bind,source="$(pwd)"/external-data.yml,target=/home/renderer/src/openstreetmap-carto/external-data.yml \
    --network=osm \
    "url to internal docker registry"/overv_openstreetmap-tile-server:1.8.2 \
    import
```

Run:
```bash
docker run \
 -p 8080:80 \
 --name openstreetmap \
 -v  "$(pwd)"/data:/var/lib/postgresql/12/main \
 -d  "url to internal docker registry"/overv_openstreetmap-tile-server:1.8.2 \
 run
```

**Comment by sevensamurai** [comment]
thanks @DigitalCop I was thinking on doing something similar, will try it out soon

**Comment by Horrih** [comment]
TL;DR : for a quick and dirty fix docker pull overv/openstreetmap-tile-server:v1.6.0

Using version 1.6.0 and before does not use those external assets.

I got in the same situation and this in indeed i real PITA. My corporate network is slow and proxy authentication only lasts for one hour, so downloading all these external assets fails miserably partway

Thank you for you guys hard work, it would be really nice to provide an easy work-around when openstreetmap carto allow loading from files (a PR seems to be in progress on their side)

This is a major issue at least for me : one of the main reasons of using your image is for people like me who cannot use one of the tile managers with a direct web access. You get to jump through many more hoops than before

**Comment by galewis2** [comment]
Here's my solution that I've used before, I download the assets into a Dockerfile and then re-host them locally:
Then as other mentioned, I change the external-data.yml to point to this.

With the changes upcoming in https://github.com/Overv/openstreetmap-tile-server/pull/243 , this should be a lot easier

```
FROM nginx:1.21.3

RUN apt-get update && apt-get -y install wget

RUN wget -P "/usr/share/nginx/html/" "https://osmdata.openstreetmap.de/download/simplified-water-polygons-split-3857.zip" \
  && wget -P "/usr/share/nginx/html/" "https://osmdata.openstreetmap.de/download/water-polygons-split-3857.zip" \
  && wget -P "/usr/share/nginx/html/" "https://osmdata.openstreetmap.de/download/antarctica-icesheet-polygons-3857.zip" \
  && wget -P "/usr/share/nginx/html/" "https://osmdata.openstreetmap.de/download/antarctica-icesheet-outlines-3857.zip" \
  && wget -P "/usr/share/nginx/html/" "https://naturalearth.s3.amazonaws.com/110m_cultural/ne_110m_admin_0_boundary_lines_land.zip" \
  && chown nginx:nginx /usr/share/nginx/html/*
  ```

**Comment by Istador** [comment]
Maybe there could be an extra optional docker tag (e.g. `v1.8.2-externals`), that has the `get-external-data.py` already executed in the last layer as part of the build process?

```Dockerfile
FROM  final  as  externals

RUN  chown  -R  renderer:renderer  /home/renderer/src  \
  &&  sudo  -E  -u renderer  python3  /home/renderer/src/openstreetmap-carto/scripts/get-external-data.py  -c /home/renderer/src/openstreetmap-carto/external-data.yml  -D /home/renderer/src/openstreetmap-carto/data
```

```bash
# build:
docker  build  --target final      --tag overv/openstreetmap-tile-server:v1.8.2            .
docker  build  --target externals  --tag overv/openstreetmap-tile-server:v1.8.2-externals  .
# deploy:
docker  push  overv/openstreetmap-tile-server:v1.8.2
docker  push  overv/openstreetmap-tile-server:v1.8.2-externals 
```

That way the download can only fail on building a new image, but not for anyone else using it.

The `run.sh` should be changed then to not execute it again, if the files already exist.

https://github.com/Overv/openstreetmap-tile-server/blob/672e2cd5b3b45ea36d9777867de4618b4c0ebe68/run.sh#L83-L85

**Comment by galewis2** [comment]
Or how about as an alternative, it's part of the multistage build, where it is the last stage?

Eg:

1. Build v1.8.2-externals
2. Build v1.8.2-core
3. Host them separately, that way it can be transported separately and doesn't massively increase the image size, especially because externals would not change much
4. Build v1.8.2, which has a final stage of just referencing the externals

**Comment by Istador** [comment]
> Or how about as an alternative, it's part of the multistage build, where it is the last stage?
> 
> Eg:
> 
> 1. Build v1.8.2-externals
> 2. Build v1.8.2-core
> 3. Host them separately, that way it can be transported separately and doesn't massively increase the image size, especially because externals would not change much
> 4. Build v1.8.2, which has a final stage of just referencing the externals

I'm not sure what you mean exactly. Maybe there's a misconception about docker layers, tags and multistage builds? The first block in my post was an additional stage for the multi stage Dockerfile (e.g. appended to the end of the existing file).

---

```
docker  build  --target final  --tag overv/openstreetmap-tile-server:v1.8.2  .
docker  push  overv/openstreetmap-tile-server:v1.8.2
```

This does exactly what the current Dockerfile would do. The `v1.8.2` tag points to the last layer in the final stage (`EXPOSE 80 5432`). It does not build any point further than that. E.g. if the `externals` stage is defined afterwards, it is not executed, and doesn't get into the image. If it is defined above the `final` stage, it might get executed (depending on how you build, buildkit for example does build all stages in parallel and ignores stages that aren't contributing to the target stage), but it won't get into the image regardeless.

---

```
docker  build  --target externals  --tag overv/openstreetmap-tile-server:v1.8.2-externals  .
docker  push  overv/openstreetmap-tile-server:v1.8.2-externals
```

This will continue where the previous build left, previous steps aren't done again (unless you use a no cache option). It adds one single layer, that contains all the new/modified files in the filesystem generated by the `get-external-data.py` script. The new tag `v1.8.2-externals` will then point to the new layer, but `v1.8.2` will still point to the previous layer.

When you push both tags, every layer is only pushed once, because they are identical. The same goes for pulling. When you pull the `v1.8.2` tag, the extra `externals` layer isn't transfered over the network at all. It does not contributing to the image size of `v1.8.2`. And if you have `v1.8.2` already present on your computer and do pull `v1.8.2-externals`, it'll only transfer the extra layer, because all other layers are already present.

---

- ...
- [layer 31: `ENTRYPOINT ["/run.sh"]`]
- [layer 32: `CMD []`]
- [layer 33: `EXPOSE 5432 80`] <--- `v1.8.2`
- [layer 34: `/bin/sh -c chown  -R  ...`] <--- `v1.8.2-externals`

**Comment by prusswan** [comment]
Supposedly local files will be supported once gravitystorm/openstreetmap-carto#4468 gets merged

I went with v1.6.0 for the time being (it takes 2hrs+ to download those external files)

**Comment by nielsole** [comment]
The above ticket did get merged in the meantime btw
